File: 2562-count-ways-to-build-good-strings/2562-count-ways-to-build-good-strings.cpp
class Solution {
public:
    long mod = 1e9 + 7, res = 0; 
    int zero, one, low, high;
    long dp[1000000];
    long helper(int size) {
        if(size > high)return 0;
        if(dp[size] != -1)return dp[size];
        int res = 0;
        if(size >= low && size <= high)
            res++;

        res += helper(size + zero);
        res += helper(size + one);
        res = res % mod;
        return dp[size] = res;
         
    }

    int countGoodStrings(int l, int h, int z, int o) {
        memset(dp, -1, sizeof(dp));
        low = l; high = h; zero = z; one = o;
        return helper(0);
    }
};

File: 2562-count-ways-to-build-good-strings/2562-count-ways-to-build-good-strings.java
class Solution {
    private int low, high, zero, one;
    private long mod = 1000000007;
    long[] dp;

    private long countGoodStringsHelper(int size) {
        if(size > high)return 0;
        if(dp[size] != -1)return dp[size];
        long res = 0;
        if(size >= low && size <= high)res++;

        res += countGoodStringsHelper(size + one);
        res = res % mod;
        res += countGoodStringsHelper(size + zero);
        res = res % mod;
        dp[size] = res;
        return res;
    }
    public int countGoodStrings(int low, int high, int zero, int one) {
        dp = new long[1000000];
        Arrays.fill(dp, -1);
        this.one = one; this.high = high; this.low = low; this.zero = zero;
        return (int) (countGoodStringsHelper(0));
        
    }
}